Added mock data for project creation

abdulimtiaz opened the pull request:
Added some mock data for project creation.

abdulimtiaz committed:
added mock data for project creation

abdulimtiaz committed:
renaming files

abdulimtiaz committed:
script to pull projects from github

abdulimtiaz commented:
test

abdulimtiaz commented:
Wrote a script to pull projects from GitHub and put them in the same format as the projects in `projects.js`.

To use:
- go to https://github.com/settings/tokens/new?scopes=repo and generate a token
- add the token to your `.env` file as `GITAUTH`
- run `npm i`
- run `node src/utils/mock_data/pullData`

saifulislamdev commented on src/utils/seedDB.js:
```diff
@@ -1,7 +1,9 @@
-const createUsersMockData = require('./mock_data/createUsersMockData');
```
The renaming of the file would have been done automatically when I merged my code in. There might be some errors once we merge both of our codes (associated with the renamed file), but no worries, I'll fix it if the issue arises.

saifulislamdev commented on src/utils/mock_data/pullData.js:
```diff
@@ -0,0 +1,126 @@
+    params.append('order', 'desc');
+    params.append('per_page', '100');
+    params.append('page', `${page}`);
+    let url = `/search/repositories?${params.toString()}`;
```
The way you wrote this code is great

saifulislamdev reviewed (changes requested):
Amazing code, can we automate the creation of the mock data without having to run a `node` command?

abdulimtiaz committed:
automated creation of mock data for projects upon starting server

abdulimtiaz commented:
Running it manually through `node` and `main()` was only for testing purposes. The latest commit automates it upon starting the server.

Other changes:
- replaced `-` character with space in the names of pulled repos and their associated topics
- added option to disable automatic mock project data creation by setting `NO_MOCK_PROJECTS` in `.env` to a non-falsy value (e.g. `123`)

abdulimtiaz committed:
changed starting id to 11

abdulimtiaz committed:
added mock project data from predefined array instead of fetching on server start

abdulimtiaz committed:
added mock data for project join requests and request handling

abdulimtiaz committed:
added .env variable check for inserting mock data

abdulimtiaz committed:
deleted unnecessary comment

abdulimtiaz commented:
- added mock data for project join requests and request handling (1 user is rejected from budcode project, others are accepted)
- renamed `projects.js` to `initialProjects.js`
- added array in `projectInfos.js` that contains the projects pulled from GitHub and inserted the data from the array
- added `.env` variable check: mock data will only be inserted if `MOCK_DATA_INSERTION=true` in `.env`

saifulislamdev commented on src/utils/mock_data/createJoinRequests.js:
```diff
@@ -0,0 +1,42 @@
+const createJoinRequests = async () => {
+    try {
+
+        const { rowCount } = await pool.query(`SELECT id FROM "ProjectJoinRequest" LIMIT 20`);
```
Great way of doing it; didn't think of doing it this way

saifulislamdev commented on src/utils/mock_data/createProjects.js:
```diff
@@ -0,0 +1,111 @@
+const createProjects = async () => {
+    try {
+
+        // check if mock data was already inserted
```
Not mock data, but data in general

saifulislamdev commented on src/utils/mock_data/handleJoinRequests.js:
```diff
@@ -0,0 +1,59 @@
+    let promises = [];
+
+    for (const joinRequest of joinRequests) {
+        if (joinRequest.username === 'patrick' && joinRequest.id === 1) {
```
Clarifying comment could be added here

saifulislamdev commented on src/utils/mock_data/initialProjects.js:
```diff
@@ -0,0 +1,146 @@
+        description: `A software development collaboration platform`,
+        skills: ['Node', 'React', 'HTML', 'CSS', 'PostgreSQL'],
+        tags: ['web development', 'full-stack development'],
+        links: [
```
I'll edit this to include a frontend GitHub repo link

saifulislamdev commented on src/utils/mock_data/initialProjects.js:
```diff
@@ -0,0 +1,146 @@
+        links: [
+            {
+                type: 'GitHub',
+                link: 'https://github.com/200ok-ch/organice'
```
Their repo is public now, so I'll change this as well

saifulislamdev commented on src/utils/mock_data/createJoinRequests.js:
```diff
@@ -0,0 +1,42 @@
+        for (const joinRequest of joinRequests) {
+            promises.push(
+                pool.query(
+                    `INSERT INTO "ProjectJoinRequest"(project_id, username, message) VALUES($1, $2, $3)`,
```
You insert into the `project_id` column the `joinRequest.id`, is this a typo? Or is `joinRequest.id` the project ID?

abdulimtiaz replied:
I should have made it more clear, but `joinRequest.id` is the project ID. I didn't manually specify the request ID because I didn't need to reference a specific request anywhere except in one spot (the one request denial) which I just referenced using the project id and username since that pairing has to be unique.

saifulislamdev commented on src/utils/mock_data/pullData.js:
```diff
@@ -0,0 +1,98 @@
+
+
+/**
+ * @param {string} query - search query
```
Where'd you learn how to document like this? I want to learn, too

abdulimtiaz replied:
VSCode has built in support for [JSDoc](https://jsdoc.app/index.html) which I thought would be useful here because documenting with it allows you to see the specified arguments and return type when you hover your cursor over the function, so you don't have to keep scrolling up and down to double check.

saifulislamdev replied:
That's amazing

saifulislamdev commented on src/utils/mock_data/pullData.js:
```diff
@@ -0,0 +1,98 @@
+ */
+const fetchRepos = async (query) => {
+
+    const octokit = new Octokit({auth: process.env.GITAUTH});
```
Do we still need to have an env for GITAUTH?

abdulimtiaz replied:
I already put the project data in the GitHub repository itself, so we shouldn't have to run the script anymore, but if anyone would want to use the script to pull fresh data, they would need it.

saifulislamdev commented on src/utils/mock_data/pullData.js:
```diff
@@ -0,0 +1,98 @@
+const generateProjects = async (topics) => {
+
+    // start at id=6 because of reserved first 5 ids
+    let id = 11;
```
The length of `initialProjects` could have been taken here instead instead of hard-coding

saifulislamdev commented on src/utils/mock_data/pullData.js:
```diff
@@ -0,0 +1,98 @@
+        repos.forEach((repo) => {
+            process.stdout.write(`\rAdding project ${id} from page ${i+1}...`);
+            let creator = usersInfos[getRandomInt(0, 20)].username;
+            projects.push(repoFormat(repo, id, creator));
```
Amazing quality code

saifulislamdev committed:
Merge branch 'main' into data/abdul-mock-data

saifulislamdev committed:
rearranged order of mock data creations to insert referenced rows first

saifulislamdev committed:
modified environment variables section of README to document new environment variables

saifulislamdev committed:
added/modified console messages to provide more insight on table and mock data creation

saifulislamdev committed:
added/modified inline documentation

saifulislamdev committed:
added another link to a project in initial projects mock data

saifulislamdev committed:
replaced hard-coded initial value for id + prettier formatting in file

saifulislamdev commented:
I made some changes in your branch, hope that's not an issue. Changes since your last commit are outlined in my commit messages. One particular thing I did was merge the main branch into this branch (your branch) to resolve the merge conflicts with my mock data code.
